script/tool/lib/src/common/xcode.dart

```diff
@@ -0,0 +1,160 @@
+
+import 'dart:convert';
+import 'dart:io' as io;
+
+import 'package:file/file.dart';
+
+import 'core.dart';
+import 'process_runner.dart';
+
+const String _xcodeBuildCommand = 'xcodebuild';
+const String _xcRunCommand = 'xcrun';
+
+/// A utility class for interacting with the installed version of Xcode.
+class Xcode {
+  /// Creates an instance that runs commends with the given [processRunner].
+  ///
+  /// If [log] is true, commands run by this instance will long various status
+  /// messages.
+  Xcode({
+    this.processRunner = const ProcessRunner(),
+    this.log = false,
+  });
+
+  /// The [ProcessRunner] used to run commands. Overridable for testing.
+  final ProcessRunner processRunner;
+
+  /// Whether or not to log when running commands.
+  final bool log;
+
+  /// Runs an `xcodebuild` in [directory] with the given parameters.
+  Future<int> runXcodeBuild(
+    Directory directory, {
+    List<String> actions = const <String>['build'],
+    required String workspace,
+    required String scheme,
+    String? configuration,
+    List<String> extraFlags = const <String>[],
+  }) {
+    final List<String> args = <String>[
+      _xcodeBuildCommand,
+      ...actions,
+      if (workspace != null) ...<String>['-workspace', workspace],
+      if (scheme != null) ...<String>['-scheme', scheme],
+      if (configuration != null) ...<String>['-configuration', configuration],
+      ...extraFlags,
+    ];
+    final String completeTestCommand = '$_xcRunCommand ${args.join(' ')}';
+    if (log) {
+      print(completeTestCommand);
+    }
+    return processRunner.runAndStream(_xcRunCommand, args,
+        workingDir: directory);
+  }
+
+  /// Returns true if [project], which should be an .xcodeproj directory,
+  /// contains a target called [target], false if it does not, and null if the
+  /// check fails (e.g., if [project] is not an Xcode project).
+  Future<bool?> projectHasTarget(Directory project, String target) async {
+    final io.ProcessResult result =
+        await processRunner.run(_xcRunCommand, <String>[
+      _xcodeBuildCommand,
+      '-list',
+      '-json',
+      '-project',
+      project.path,
+    ]);
+    if (result.exitCode != 0) {
+      return null;
+    }
+    Map<String, dynamic>? projectInfo;
+    try {
+      projectInfo = (jsonDecode(result.stdout as String)
+          as Map<String, dynamic>)['project'] as Map<String, dynamic>?;
+    } on FormatException {
+      return null;
+    }
+    if (projectInfo == null) {
+      return null;
+    }
+    final List<String>? targets =
+        (projectInfo['targets'] as List<dynamic>?)?.cast<String>();
+    return targets?.contains(target) ?? false;
+  }
+
+  /// Returns the newest available simulator (highest OS version, with ties
+  /// broken in favor of newest device), if any.
+  Future<String?> findBestAvailableIphoneSimulator() async {
+    final List<String> findSimulatorsArguments = <String>[
+      'simctl',
+      'list',
+      '--json'
+    ];
```

Comment:
I know you just copied this, but if you add `devices available`
```suggestion
    final List<String> findSimulatorsArguments = <String>[
      'simctl',
      'list',
      'devices',
      'available',
      '--json'
    ];
```
The `availabilityError`/`isAvailable` check below can be removed.

Reply:
Thanks, done! (I had to add 'runtimes' as well since it's parsing those too.)